### PR TITLE
106 gradient hmc clock

### PR DESCRIPTION
**Pull request by christiaanjs**
Includes implementation of branch length gradients and custom conda recipe for BEAGLE branch

### CODE REVIEW

File: beagle-recipe/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+build:
+  number: 1
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - autoconf
+    - automake
+    - libtool
+    - openjdk
```
Comment (matsen) [inline beagle-recipe/meta.yaml]: It always seemed strange to me that the beagle conda required openjdk. Have you tried without?

File: src/beagle.cpp [hunk 2]
```diff
@@ -388,4 +408,125 @@ std::vector<std::pair<double, std::vector<double>>> BranchGradients(
...
+  for (int index : derivative_matrix_indices) {
+    beagleSetDifferentialMatrix(beagle_instance, index, Q.data());
+  }
+
+  beagleUpdatePartials(beagle_instance, operations.data(),
+                       static_cast<int>(operations.size()),
+                       BEAGLE_OP_NONE);  // cumulative scale index
+
+  // Calculate pre-order partials
+  // TODO: Set root pre-order partials equal to root state frequencies
```
Comment (matsen) [inline src/beagle.cpp]: Is this still outstanding?

File: beagle-recipe/meta.yaml (hunk 1 above)
Comment (christiaanjs) [reply beagle-recipe/meta.yaml]: It looks like a JDK is an optional dependency for Beagle so I will remove it

**Comment by christiaanjs** [comment]
The new branch length gradient implementation is forked from @4ment . Should I go ahead and replace the existing implementation with it/tidy it up or leave it to him? Also, it doesn't look like I have push permissions to the `phylovi/libsbn` repository anymore.

**Comment by christiaanjs** [comment]
Also, I've added the setup process to the Dockerfile but I'm not very familiar with it so I'm not sure how to get the Travis build to pass

**Comment by matsen** [comment]
I just chatted with @4ment who suggests that you replace the existing implementation.

Regarding the Travis process, you should be able to simulate it by running 

    docker build -t phylovi/libsbn .

then 

    docker run -v ${PWD}:/libsbn -t phylovi/libsbn /bin/bash -c "/opt/conda/bin/conda run -n libsbn make -C /libsbn"

**Pull request closed without merging**